crates/resolver/src/lookup_ip.rs

```diff
@@ -317,6 +330,74 @@ impl<C: DnsHandle> LookupContext<C> {
     }
 }
 
+pin_project! {
+    struct RacingLookup<A, Aaaa, F> {
+        #[pin]
+        a: A,
+        a_output: Option<Result<Lookup, ProtoError>>,
+        #[pin]
+        aaaa: Aaaa,
+        aaaa_output: Option<Result<Lookup, ProtoError>>,
+        delay: Option<F>,
+        sleeping: Option<Pin<Box<dyn Future<Output = ()> + Send>>>,
+    }
+}
+
+impl<A, Aaaa, F> Future for RacingLookup<A, Aaaa, F>
+where
+    A: Future<Output = Result<Lookup, ProtoError>>,
+    Aaaa: Future<Output = Result<Lookup, ProtoError>>,
+    F: FnOnce() -> Pin<Box<dyn Future<Output = ()> + Send>> + Send,
+{
+    type Output = Result<Lookup, ProtoError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        // Kick off the AAAA query first
+        if this.aaaa_output.is_none() {
+            match this.aaaa.poll(cx) {
+                Poll::Ready(Ok(aaaa)) if !aaaa.is_empty() => return Poll::Ready(Ok(aaaa)),
+                Poll::Ready(res @ Err(_) | res @ Ok(_)) => match this.a_output.take() {
+                    None => *this.aaaa_output = Some(res),
+                    Some(Ok(a)) if !a.is_empty() => return Poll::Ready(Ok(a.clone())),
+                    Some(_) => return Poll::Ready(res),
+                },
+                Poll::Pending => {}
+            }
+        }
+
+        if this.a_output.is_none() {
+            match this.a.poll(cx) {
+                Poll::Ready(Ok(a)) if !a.is_empty() => {
+                    if let Some(sleep) = this.delay.take() {
+                        *this.a_output = Some(Ok(a));
+                        *this.sleeping = Some((sleep)());
+                    }
+                }
+                Poll::Ready(res) => {
+                    *this.a_output = Some(res);
+                    if let Some(res) = this.aaaa_output.take() {
```

Comment:
I think it would be good to document this case, it appears to be saying, if A has a none-positive result, and there is also a AAAA result, then return the AAAA. Which I think is correct, but it's not obvious on first read-through.